### PR TITLE
Add ensure automatic scenario assigment

### PR DESCRIPTION
**Pull request by dbadura**
**Description**

Changes proposed in this pull request:

- Add query which updates scenarios for runtimes
- Unify database configuration

**Related issue(s)**

See #1031

### CODE REVIEW

File: components/director/internal/domain/scenarioassignment/service_test.go [hunk 1]
```diff
@@ -354,7 +356,7 @@ func TestService_DeleteManyForSameSelector(t *testing.T) {
 		// GIVEN
 		mockRepo := &automock.Repository{}
 		mockRepo.On("DeleteForSelector", ctx, tenantID, selector).Return(nil).Once()
-		mockEngineSvc := &automock.AssignmentEngineService{}
+		mockEngineSvc := &automock.AssignmentEngine{}
```
Comment (dbadura) [inline components/director/internal/domain/scenarioassignment/service_test.go:359]: What is the best name:
`mockEngine` or `mockEngineSvc`. I would go with `mockEnginge.

File: tests/director/go.mod [hunk 2]
```diff
@@ -22,3 +22,5 @@ require (
 	golang.org/x/sys v0.0.0-20200212091648-12a6c2dcc1e4 // indirect
 	k8s.io/apimachinery v0.17.3 // indirect
 )
+
+replace github.com/kyma-incubator/compass/components/director v0.0.0-20200403141306-1c5459e649a7 => github.com/dbadura/compass/components/director v0.0.0-20200403132654-f75f8e11a801
```
Comment (dbadura) [inline tests/director/go.mod:26]: I will remove this after merge.

File: components/director/internal/domain/scenarioassignment/engine.go [hunk 3]
```diff
@@ -1,17 +1,39 @@
...
+			AND l.key ='scenarios'
+			AND l.tenant_id=$3`
+
+func (e *engine) EnsureScenarioAssigned(ctx context.Context, in model.AutomaticScenarioAssignment, tenantID string) error {
+	persist, err := persistence.FromCtx(ctx)
+	if err != nil {
+		return errors.Wrap(err, "while getting persitance from context")
+	}
+
+	_, err = persist.Exec(updateQuery, in.Selector.Key, in.Selector.Value, tenantID)
```
Comment (kjaksik) [inline components/director/internal/domain/scenarioassignment/engine.go:32]: Please move the db operations to some kind of a repository

File: tests/director/api/fixtures_queries_test.go [hunk 4]
```diff
@@ -10,6 +10,21 @@ import (
...
 )
 
+func SetAutomaticScenarioAssignmentWithinTenant(t *testing.T, ctx context.Context, tenantID, automaticScenario, selecterKey, selectorValue string) {
+	input := graphql.AutomaticScenarioAssignmentSetInput{
+		ScenarioName: automaticScenario,
+		Selector: &graphql.LabelSelectorInput{
+			Key:   selecterKey,
+			Value: selectorValue,
+		},
+	}
```
Comment (kjaksik) [inline tests/director/api/fixtures_queries_test.go:20]: Hope it won't break go fmt
```suggestion
func SetAutomaticScenarioAssignmentWithinTenant(t *testing.T, ctx context.Context, tenantID, automaticScenario, selectorKey, selectorValue string) {
	input := graphql.AutomaticScenarioAssignmentSetInput{
		ScenarioName: automaticScenario,
		Selector: &graphql.LabelSelectorInput{
			Key:   selectorKey,
			Value: selectorValue,
		},
	}
```

File: tests/director/api/fixtures_queries_test.go [hunk 5]
```diff
@@ -10,6 +10,21 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func SetAutomaticScenarioAssignmentWithinTenant(t *testing.T, ctx context.Context, tenantID, automaticScenario, selecterKey, selectorValue string) {
```
Comment (kjaksik) [inline tests/director/api/fixtures_queries_test.go:13]: Apart from the typo we already have 2(???) fixtures with  similar functionality, please decide which is the best one and clean up the rest.

Comment (kjaksik) [reply tests/director/api/fixtures_queries_test.go:13]: [Here](https://github.com/kyma-incubator/compass/blob/86d9878d6bb98a35e05a1b21aeb6a081a7999475/tests/director/api/fixtures_queries_test.go#L461) and [here](https://github.com/kyma-incubator/compass/blob/86d9878d6bb98a35e05a1b21aeb6a081a7999475/tests/director/api/fixtures_queries_test.go#L472)

File: tests/director/api/automatic_scenario_assignment_test.go [hunk 6]
```diff
@@ -82,7 +82,57 @@ func TestAutomaticScenarioAssignmentQueries(t *testing.T) {
...
+func Test_AutomaticScenarioAssigmentForRuntime(t *testing.T) {
+	//GIVEN
+	ctx := context.TODO()
+	tenantID := testTenants.GetIDByName(t, "TestCreateAutomaticScenarioAssignment")
+	prodScenario := "PRODUCTION"
+	devScenario := "DEVELOPMENT"
+	createScenariosLabelDefinitionWithinTenant(t, ctx, tenantID, []string{prodScenario, devScenario, "DEFAULT"})
+
+	rmtInput := fixRuntimeInput("runtime1")
+	rtm1 := registerRuntimeFromInputWithinTenant(t, ctx, &rmtInput, tenantID)
```
Comment (dbadura) [inline tests/director/api/automatic_scenario_assignment_test.go:96]: I would rename those magic names

File: components/director/internal/domain/scenarioassignment/repository.go [hunk 7]
```diff
@@ -127,3 +137,16 @@ func (r *repository) DeleteForScenarioName(ctx context.Context, tenantID string,
 
 	return r.deleter.DeleteOne(ctx, tenantID, conditions)
 }
+
+func (r *repository) EnsureScenarioAssigned(ctx context.Context, in model.AutomaticScenarioAssignment) error {
```
Comment (pkosiec) [inline components/director/internal/domain/scenarioassignment/repository.go:141]: I tested it and it doesn't work right way. Basically it removes all other scenarios for given runtimes. It should preserve them and just append a new one from ASA.

- Add new scenario
``` graphql
mutation addNewScenario {
  createLabelDefinition(in: {
    key:"scenarios",
    schema: "{\"items\":{\"enum\":[\"DEFAULT\", \"FOO\", \"BAR\"],\"maxLength\":128,\"pattern\":\"^[A-Za-z0-9]([-_A-Za-z0-9\\\\s]*[A-Za-z0-9])$\",\"type\":\"string\"},\"minItems\":1,\"type\":\"array\",\"uniqueItems\":true}"  }) {
    key
      schema
  }
}
```
- Create runtimes

```graphql

mutation CreateRTM1 {
  registerRuntime(in:{
    name:"foo",
    labels: {
      foo:"bar",
      scenarios: ["DEFAULT", "FOO"],
    }
  }) {
    id
    name
  }
}

mutation CreateRTM2 {
  registerRuntime(in:{
    name:"bar",
    labels: {
      foo:"bar",
      scenarios: ["DEFAULT"],
    }
  }) {
    id
    name
  }
}

mutation CreateRTM3 {
  registerRuntime(in:{
    name:"baz",
    labels: {
      foo:"bar",
      scenarios: ["DEFAULT", "BAR"],
    }
  }) {
    id
    name
  }
}
```

- Create ASA

```graphql
mutation CreateASA {
  createAutomaticScenarioAssignment(in:{
    scenarioName:"FOO",
    selector:{
      key:"foo",
      value:"bar"
    }
  }) {
    scenarioName
    selector {
      key
      value
    }
  }
}
```

- List Runtimes:
```graphql
  runtimes {
    data {
      id
      name
      labels
    }
  }
```

Observe that both runtimes have just `FOO` scenario assigned:

```json
  {
  "data": {
    "runtimes": {
      "data": [
        {
          "id": "f1fec278-37bb-4d16-a61b-611968e46c09",
          "name": "bar",
          "labels": {
            "foo": "bar",
            "scenarios": [
              "FOO"
            ]
          }
        },
        {
          "id": "63b35c2c-65c8-49e8-91cf-f167a0f72447",
          "name": "baz",
          "labels": {
            "foo": "bar",
            "scenarios": [
              "FOO"
            ]
          }
        },
        {
          "id": "e91385d5-59a9-46a0-8e30-c11f2abaa09e",
          "name": "foo",
          "labels": {
            "foo": "bar",
            "scenarios": [
              "FOO"
            ]
          }
        }
      ]
    }
  }
}
```

File: components/director/pkg/persistence/config.go [hunk 8]
```diff
@@ -9,12 +9,12 @@ type DatabaseConfig struct {
 	Password           string `envconfig:"default=pgsql@12345,APP_DB_PASSWORD"`
 	Host               string `envconfig:"default=localhost,APP_DB_HOST"`
 	Port               string `envconfig:"default=5432,APP_DB_PORT"`
-	Name               string `envconfig:"default=postgres,APP_DB_NAME"`
+	DBName             string `envconfig:"default=compass,APP_DB_NAME"`
```
[the remark this answers is not in the document]
Comment (dbadura) [reply components/director/pkg/persistence/config.go:12]: I agree, but everytime I read `Name` forgot the  meaning.
I will revert this change :)

File: components/director/internal/domain/scenarioassignment/engine.go [hunk 9]
```diff
@@ -1,17 +1,68 @@
...
+		output, ok := scenario.(string)
+		if !ok {
+			return nil, errors.New("scenario is not a string")
+		}
+		set[output] = struct{}{}
+	}
+
+	return str.MapToSlice(set), nil
+}
+func (e *engine) mergeScenarios(ctx context.Context, labels []model.Label, scenarioName string, mergeFn func(scenarios []interface{}, diffScenario string) ([]string, error)) error {
```
Comment (pkosiec) [inline components/director/internal/domain/scenarioassignment/engine.go:49]: What do you think about name `upsertMergedScenarios`?

**Pull request merged**